pageserver/src/tenant_mgr.rs

```diff
@@ -204,36 +151,27 @@ pub fn create_repository_for_tenant(
     conf: &'static PageServerConf,
     tenantid: ZTenantId,
 ) -> Result<()> {
-    {
-        let mut m = access_tenants();
-        // First check that the tenant doesn't exist already
-        if m.get(&tenantid).is_some() {
-            bail!("tenant {} already exists", tenantid);
+    let wal_redo_manager = Arc::new(PostgresRedoManager::new(conf, tenantid));
+    let repo = Some(branches::create_repo(conf, tenantid, wal_redo_manager)?);
+
+    match access_tenants().entry(tenantid) {
+        hash_map::Entry::Occupied(_) => bail!("tenant {} already exists", tenantid),
+        hash_map::Entry::Vacant(v) => {
+            v.insert(Tenant {
+                state: TenantState::Idle,
+                repo,
+            });
         }
-        let tenant = Tenant {
-            state: TenantState::CloudOnly,
-            repo: None,
-        };
-        m.insert(tenantid, tenant);
     }
 
-    let wal_redo_manager = Arc::new(PostgresRedoManager::new(conf, tenantid));
-    let repo = branches::create_repo(conf, tenantid, wal_redo_manager)?;
-
-    let mut m = access_tenants();
-    let tenant = m.get_mut(&tenantid).unwrap();
-    tenant.repo = Some(repo);
-    tenant.state = TenantState::Idle;
-
     Ok(())
 }
 
-// If tenant is not found in the repository, return CloudOnly state
+// If tenant is not found in the repository, return Stopping state
```

Comment:
Could you please elaborate, why do we treat missing tenant as Stopping?

Reply:
It's not my code originally, but I've removed the `CloudOnly` tenant state, since it turned to be totally unrelated to any s3 sync code.

Maybe `TenantState` is misleading? Right now, this enum seems to be mostly used to indicate that a WAL receiver is enabled for this tenant (ergo needs gc and related activities).
Old `CloudOnly` in the old code was just `Ignore` semantically.

I'll change this function to return `Option`, since I agree it's a callers responsibility to understand what status to use (if any) in exceptional cases.